### DISCUSSION TITLE
Extend ui.number widget

### DISCUSSION THREAD

**Discussion by 1081** [zauberzeug/nicegui#765] [Ideas / Feature Requests]
First great work with nicegui! 

I like to use nicegui to build tools for scientific calculations.
- Would it be possible to specify the amount of the spinner for the widget. e.g. jump 100 instead of 1 when pushing the up button.
- Specify min and max values for the widget
- Add fixed suffix-string to the number to indicate the unit next to the number. So if the input is "123" it shows "123 mm" . This would look very nice and professional if many of number widgets are stacked vertically. 
- Set the precision of the input number, if precision is set to 0.1 the widget value should be rounded to that precision.

**Comment by falkoschindler** [comment]
Thanks for these suggestions, @1081!
Most of it is already possible:

```py
ui.number('Percentage', value=100, format='%.1f').props('min=50 max=150 step=10 suffix=%').classes('w-20')
```

But we could provide a more official and better documented API, introducing `min`, `max`, `step` and `suffix` parameters.

On remaining issue might be, that the format only affects the displayed value. But if you, e.g., input "123.456", the displayed value might be rounded to "123.5", but the value used for computations is still "123.456". Do you think this is a problem?

**Reply by falkoschindler** [reply]
By the way: Introducing those new parameters might be a good opportunity for a pull request to become an official contributor. 😀

**Reply by 1081** [reply]
I would love to contribute to this awesome project, but am still very new to it. :-)

**Comment by 1081** [comment]
hi Falko,
sounds great, the documentation for the .props() would be very helpful.

Regarding the display format, yes this might be a problem. At least for my kind of application (numerical calculations) this behavior is dangerous. Calculating with different values than displayed is a source of errors.

I have the requirement to define or better limit the precision of the input.
For example, if I require a maximal decimal precision of 3 and put in the value 0,0114 by mistake, the widget shows 0,011 but the calculation does 0,0114**2.

**Reply by falkoschindler** [reply]
Ok, forcing the value to respect the display format can be easily added with a change handler like this:
```py
ui.number(format='%.1f', on_change=lambda e: e.sender.set_value(e.sender.format % e.value))
```
Now I'm thinking about a name for a built-in parameter that automatically does this formatting.
Or should we change the current behavior such that the displayed value is _always_ the true value? First I thought one might want to input more digits than displayed. But on the other hand it doesn't make much sense if the UI element expects fewer digits. Like a `ui.slider` with a certain step size a `ui.number` could also limit the input to a certain level of granularity.

**Comment by falkoschindler** [comment]
I created PR https://github.com/zauberzeug/nicegui/pull/790 implementing all discussed features. 🙂
